### ISSUE TITLE
Allow specific localhost resources to be accessible by all websites.

### ISSUE THREAD

**Issue by MicahZoltu** [brave/brave-browser#30181]
## Description 

Support for permissioned access to localhost was added in https://github.com/brave/brave-core/pull/17321, but by default sites cannot request this permission.  They must first be manually added to a whitelist, and at that point the website can request access to *all* localhost resources.

Sometimes there are locally available resources that one would like all websites to be able to access.  In particular, local resources that facilitate a more decentralized web like IPFS or Ethereum.  There should be a way for users to add a specific localhost resource (port) such that any website can request permission to access just that resource.  For example, you might add `localhost:8080` (IPFS default port) and `localhost:8545` (Ethereum default HTTP port) with a label (IPFS and Ethereum respectively) and any website would be able request access to Local IPFS or request access to Local Ethereum the same way they request access to camera/mic.

Generally speaking, I think there are very few situations where there is a good reason to allow a website to port scan the local machine, but I can imagine a number of reasons (integrating with desktop services) where access to a very specific port is quite reasonable for any website that wants to integrate with it.

----

Note: Opening this per request from @ShivanKaul over in https://github.com/brave/brave-browser/issues/27346#issuecomment-1536845368.  I didn't see an option for opening a feature request, let me know if this is the wrong place.

**Comment by ShivanKaul** [comment]
There is a way to say "allow requests to localhost resources by all websites", but not "allow requests to this particular localhost port by any/all websites". 

![image](https://github.com/brave/brave-browser/assets/5284154/6866b9be-f9c9-4f1b-8d90-11695b8e4d17)

**Comment by ShivanKaul** [comment]
It might be useful for this permission to key off of the localhost port as well as the embedding origin.

**Comment by MicahZoltu** [comment]
I think it is also important to allow the "labeling" of ports.  I can imagine in the future there being a way for external applications to add/suggest entries into this list (e.g., when you install a desktop app that exposes a localhost bound port) and being able to label it would make it so when a permission request occurs for a specific port Brave would tell the user the name rather than the number.

For example, "www.some-website.com wants to access IPFS from your computer, allow/deny" vs "www.some-website.com wants to access port 8080 from your computer, allow/deny".  This would be similar to the labeling of microphones/cameras rather than "USB device #3".

**Comment by MicahZoltu** [comment]
@ShivanKaul I don't believe this is a duplicate of that issue.  This feature request is still relevant even when using Chrome's implementation as Chrome's implementation has the same privacy/security/fingerprinting problem as Brave's original implementation.

**Comment by ShivanKaul** [comment]
> This feature request is still relevant even when using Chrome's implementation as Chrome's implementation has the same privacy/security/fingerprinting problem as Brave's original implementation.

Which is?

**Comment by MicahZoltu** [comment]
In the current specification, if you grant permission to access local resources, the website can port scan your computer.  If you grant permission to access local network, the website can port scan your entire local network.  This can lead to websites seeking out vulnerable devices on the network beyond what the user intended to connect to, and it can be used to fingerprint the user.

A website can ask for either of these permissions under the guise of a legitimate request for local device/network access, but then use the too-broad permission grant to do more than just what the user was intending to allow.

There is an open discussion (which I have commented on) on the spec repository that at least tries to address the localhost side of the problem, but I am quite hesitant to rely on Google to address privacy concerns with the specification (maybe they can be swayed by the security concerns).
